From fb80a19d7daed92a05a97563a05e15924d0d9b12 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Thu, 18 Sep 2025 15:48:30 -0500
Subject: [PATCH 309/378] Revert "HACK: winex11: Ignore KWin-originated
 maximized window state changes."

This reverts commit dfe45c50b5f71eb3827fe2277f35a2b75d828608.
---
 dlls/winex11.drv/window.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index bf3a17fbefb..e9b74aba75e 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -1846,9 +1846,10 @@ static UINT window_update_client_state( struct x11drv_win_data *data )
     if (data->current_state.net_wm_state & (1 << NET_WM_STATE_MAXIMIZED)) new_style |= WS_MAXIMIZE;
 
     /* KWin sometimes combines NET_WM_STATE_FULLSCREEN with NET_WM_STATE_MAXIMIZED, but sometimes doesn't.
-     * Don't feed back the maximized state to the Win32 side as it confuses many applications.
+     * Don't feeding back the maximized state to the Win32 side when windows are fullscreen.
      */
-    if (X11DRV_HasWindowManager( "KWin" )) new_style = (new_style & ~WS_MAXIMIZE) | (old_style & WS_MAXIMIZE);
+    if (X11DRV_HasWindowManager( "KWin" ) && data->current_state.net_wm_state & (1 << NET_WM_STATE_FULLSCREEN))
+        new_style = (new_style & ~WS_MAXIMIZE) | (old_style & WS_MAXIMIZE);
 
     if ((old_style & WS_MINIMIZE) && !(new_style & WS_MINIMIZE))
     {
@@ -1913,9 +1914,10 @@ static UINT window_update_client_config( struct x11drv_win_data *data )
     if (data->current_state.net_wm_state & (1 << NET_WM_STATE_MAXIMIZED)) new_style |= WS_MAXIMIZE;
 
     /* KWin sometimes combines NET_WM_STATE_FULLSCREEN with NET_WM_STATE_MAXIMIZED, but sometimes doesn't.
-     * Don't feed back the maximized state to the Win32 side as it confuses many applications.
+     * Don't feeding back the maximized state to the Win32 side when windows are fullscreen.
      */
-    if (X11DRV_HasWindowManager( "KWin" )) new_style = (new_style & ~WS_MAXIMIZE) | (old_style & WS_MAXIMIZE);
+    if (X11DRV_HasWindowManager( "KWin" ) && data->current_state.net_wm_state & (1 << NET_WM_STATE_FULLSCREEN))
+        new_style = (new_style & ~WS_MAXIMIZE) | (old_style & WS_MAXIMIZE);
 
     if ((old_style & WS_CAPTION) == WS_CAPTION || !data->is_fullscreen)
     {
-- 
2.51.0

